From 9944f905eb1e177d3fb8d0a07153693c583d2750 Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Tue, 11 Feb 2025 10:25:11 +0800
Subject: [PATCH 01/16] i3c: master: svc: fix tx issue for SVC I3C IP 1.0
 version

There is an HW issue in SVC I3C IP 1.0:
- I3C HW stalls the write transfer if the transmit FIFO becomes empty,
when new data is written to FIFO, I3C HW resumes the transfer but
the first transmitted data bit may have the wrong value.

Workaround:
- Fill the FIFO in advance to prevent FIFO from becoming empty.

Upstream-Status: Pending [Not submitted to upstream yet]
Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 drivers/i3c/master/svc-i3c-master.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/i3c/master/svc-i3c-master.c b/drivers/i3c/master/svc-i3c-master.c
index c5ab39f1e755..fdd2582164d2 100644
--- a/drivers/i3c/master/svc-i3c-master.c
+++ b/drivers/i3c/master/svc-i3c-master.c
@@ -102,6 +102,9 @@
 #define   SVC_I3C_MDATACTRL_TXTRIG_FIFO_NOT_FULL GENMASK(5, 4)
 #define   SVC_I3C_MDATACTRL_RXTRIG_FIFO_NOT_EMPTY 0
 #define   SVC_I3C_MDATACTRL_RXCOUNT(x) FIELD_GET(GENMASK(28, 24), (x))
+#if defined(CONFIG_ARCH_NPCM)
+#define   SVC_I3C_MDATACTRL_TXCOUNT(x) FIELD_GET(GENMASK(20, 16), (x))
+#endif
 #define   SVC_I3C_MDATACTRL_TXFULL BIT(30)
 #define   SVC_I3C_MDATACTRL_RXEMPTY BIT(31)
 
@@ -1121,6 +1124,29 @@ static int svc_i3c_master_xfer(struct svc_i3c_master *master,
 	int retry = 2;
 	u32 reg;
 	int ret;
+#if defined(CONFIG_ARCH_NPCM)
+	int i, count, space;
+	/*
+	 * HW issue:
+	 * I3C HW stalls the write transfer if the transmit FIFO becomes empty,
+	 * when new data is written to FIFO, I3C HW resumes the transfer but
+	 * the first transmitted data bit may have the wrong value.
+	 * Workaround:
+	 * Fill the FIFO in advance to prevent FIFO from becoming empty.
+	 */
+	if (!rnw && xfer_len) {
+		reg = readl(master->regs + SVC_I3C_MDATACTRL);
+		space = SVC_I3C_FIFO_SIZE - SVC_I3C_MDATACTRL_TXCOUNT(reg);
+		count = xfer_len > space ? space : xfer_len;
+		for (i = 0; i < count; i++) {
+			if (i == xfer_len - 1)
+				writel(out[0], master->regs + SVC_I3C_MWDATABE);
+			else
+				writel(out[0], master->regs + SVC_I3C_MWDATAB);
+			out++;
+		}
+	}
+#endif
 
 	/* clean SVC_I3C_MINT_IBIWON w1c bits */
 	writel(SVC_I3C_MINT_IBIWON, master->regs + SVC_I3C_MSTATUS);
-- 
2.43.0

